.github/workflows/gpu-ci.yml

```diff
@@ -0,0 +1,69 @@
+name: Array API
+on:
+  workflow_dispatch:
+    inputs:
+      pr_id:
+
+  # XXX Temporary addition to execute workflow on pull requests, but only those
+  # that modify this file
+  # Remove before merging
+  pull_request:
+    paths:
+      - ".github/workflows/gpu-ci.yml"
+
+jobs:
+  tests:
+    runs-on: GPU
+    name: Run Array API unit tests
+    steps:
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.12'
+      # change this to checkout a particular PR if a PR number is provided as input
+      - name: Checkout main repository
+        uses: actions/checkout@v2
+      - name: Checkout a particular Pull Request
+        if: inputs.pr_id
+        env:
+          PR_ID: ${{ inputs.pr_id }}
+        run: |
+          echo "Checking out" ${{ env.PR_ID }}
+          gh pr checkout ${{ env.PR_ID }}
+      - name: Cache conda environment
+        id: cache-conda
+        uses: actions/cache@v3
+        env:
+          cache-name: cache-conda-env
+        with:
+          path: ~/conda
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('build_tools/github/pylatest_conda_forge_mkl_array-api_linux-64_conda.lock') }}
+      - name: Install miniforge
+        if: ${{ steps.cache-conda.outputs.cache-hit != 'true' }}
+        run: |
+          curl -L -O "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh"
```

Comment:
Is there a good reason not to use the `setup-miniconda` action: https://github.com/conda-incubator/setup-miniconda?

Reply:
I'll investigate it. I didn't use it because I didn't know it existed!

One potential complication is that conda is already setup in the VM image. However it is owned by `root` and we run as UID 1001. And `sudo chown 1001 -R ..` can not change the ownership of the existing conda install. I am not sure I understand how that is possible, but yeah.

Reply:
How important is using the action to you? I looked at it a bit and it doesn't really save us that much typing/config, but I find it harder to understand what it really does (several typescript files to follow compared to a few lines of shell) :-/

Reply:
I am with you about finding shell easier to follow than multiple typescript files but I would say that this kind of commonly used actions just work out of the box, and you very rarely have to read the typescript files to understand what it does.

For example, it is very likely that using the action would have saved you the hassle of figuring out your `chown` issue.

Reply:
I am leaning towards not using the action. It seems by the time that we also configure caching it is as complex or more complex to use the action. I think this is because we already have a lock file, so really we want to save time downloading things and there is nothing to be gained regarding solving.

Reply:
My understanding is that the action uses the available conda on a Github worker so no need to download anything and setup a cache, see https://github.com/conda-incubator/setup-miniconda?tab=readme-ov-file#conda-incubatorsetup-miniconda

> locating the conda installation [bundled](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-software) with the available runners and available in $CONDA

Reply:
If it does that, then I think we can't use it :-/ Because the conda that is installed (in the image the GPU runner uses) doesn't allow us to create new environments or install things in the existing one (because the permissions are wrong and I can't figure out how to change them)

Reply:
Ah right I did not understand that there was an issue with the GPU Github runner image which apparently is "Ubuntu NVIDIA GPU-Optimized Image for AI and HPC" looking at https://github.com/organizations/scikit-learn/settings/actions/github-hosted-runners/4?viewing_from_runner_group=false.

Fine for doing the manual conda install and revisit later.

Looking at the [build log](https://github.com/scikit-learn/scikit-learn/actions/runs/9286731267/job/25554084701?pr=29130), it looks like the error is that you can not install into the `base` environment (and we do this for installing `conda-lock`:

```
EnvironmentNotWritableError: The current user does not have write permissions to the target environment.
  environment location: /opt/miniconda
  uid: 1001
  gid: 100
```

It looks like in Azure we have the "Take ownership of conda installation" maybe we should do something similar here (`CONDA` env var may not exist in the Github image and may need to be computed by other means)?
```
sudo chown -R $USER $CONDA
```

Reply:
Thinking about it I still kind of think the setup-miniconda may just work because it will create an environment first? Needs to be double-checked of course ...

Reply:
> My understanding is that the action uses the available conda on a Github worker so no need to download anything and setup a cache (...)

But would it be possible to cache the full `conda-lock` resolution with `setup-conda`?

![image](https://github.com/scikit-learn/scikit-learn/assets/89061/a9eccfbb-d278-44d4-84a1-00bd571ef196)

It might still be worth a try to compare setup time and config verbosity of each method. Since `actions/setup-conda` is officially maintained by GitHub developers, I have the feeling that it might help us run things more smoothly/efficiently in the long term, but if it makes caching the conda-lock resolution more complex, maybe it's not worth the change.

Reply:
> But would it be possible to cache the full conda-lock resolution with setup-conda?

I don't think so, I missed the fact that this PR was caching the full conda environment. 

What I discovered though is that because we are using explicit lock-files, conda can install them directly with something like (i.e. no need to install conda-lock):
```
conda create --name test-env --file build_tools/github/pylatest_conda_forge_mkl_array-api_linux-64_conda.lock
```

so in principle you can use them directly in the `setup-miniconda` action, maybe? Since conda-lock downloads all packages I am not sure there would be such a difference with downloading from the cache but this needs to be double-checked.

Reply:
> `sudo chown -R $USER $CONDA`

I tried this but this results in an error. Which I think means in the GPU VM image the creators did something special, maybe with the use of additional linux ACL stuff.